### ISSUE TITLE
improved folder structure for generators

### ISSUE THREAD

**Issue by maximelucas** [ComplexGroupInteractions/xgi#319]
Right now, our `generators` are divided into 3 files: `classic`, `uniform`, `nonuniform`.

I see a few disadvantages. In `classic`, I feel like we've started putting everything that's not "purely random".  Also, our structure does not seem very scalable if we have more and more generators in the future. Finally, some generators are almost duplicates between `uniform` and `nonuniform` (`uniform_erdos_renyi_hypergraph` and `random_hypergraph`).

Networkx has the following structure: https://github.com/networkx/networkx/tree/main/networkx/generators
with more files (probably too many for us) with more explicit names.

We could have files like: 
- `classic`: `empty_hypergraph`, `complete_hypergraph` (new), `trivial_hypergraph`(new), etc.
- `simple`: `star_clique`, `sunflower`, etc.
- `lattice`: `ring_lattice`, + other new lattices
- `simplicial_complexes`: `flag_complex`, `flag_complex_d2`, etc. (this name might not be the best)
- `random`: `random_hypergraph`, `chung_lu_hypergraph `, `dcsbm_hypergraph `, `watts_strogatz_hypergraph`, etc.

For generators that are currently "duplicated" between `nonuniform` and `uniform`, we could could have common interface:
```python
xgi.random_hypergraph(N=10, ps=[0.1, 0.01])  # builds hyperedges for orders from 1 to len(ps)
xgi.random_hypergraph(N=10, ps=[0.1, 0.01], orders=[1, 2])  # equivalent to above
xgi.random_hypergraph(N=10, ps=[0.1])  # builds hyperedges for order 1
xgi.random_hypergraph(N=10, ps=[0.01], orders=[2])  # builds hyperedges for order 2 (uniform)
```
Other option: argument `order` that if `None` infers orders like now, and if an `int` say 2, goes uniform. I guess we would rarely need to use `orders=[2,3]` anyway.

I'm not sure if that would make sense for all generators currently in `uniform`, e.g. `configuration_model` or `planted_partition`. @nwlandry do you think we could fit those in this new structure in a way that makes sense?

This suggestion is also in line with making our file structure clearer and more discoverable.

**Comment by nwlandry** [comment]
I think that we should address #292 when we work on this.

I think that most of these suggestions make sense. The only reason that I differentiated between uniform and non-uniform is because of the Chung-Lu/DCSBM where edge sizes are not guaranteed.

**Comment by maximelucas** [comment]
Agreed on #292. 

> The only reason that I differentiated between uniform and non-uniform is because of the Chung-Lu/DCSBM where edge sizes are not guaranteed.
How do you they could fit in this then?

We could also have a `uniform` file for functions that don't have a nonuniform equivalent.

**Comment by nwlandry** [comment]
I would also say that along these lines, we should split the `matrix.py` file in `linalg` into smaller chunks. (laplacian, adjacency, etc.)

**Comment by leotrs** [comment]
One aspect I find useful when determining when to split files is when two files are usually modified at the same time, that's a good reason to actually merge them in the same file. So, conversely, if we decide to split certain files, we should take care to not fall into this pattern of always having to modify two files at the same time.

I think separating Laplacians from Adjacencies makes sense. Unifying all the random generators also makes sense.

**Comment by maximelucas** [comment]
I'll start a draft PR

**Comment by nwlandry** [comment]
Fixed with #321.
